Proposal.md

```diff
@@ -0,0 +1,604 @@
+    ///
+    /// - Complexity: O(*n*) where *n* is the length of the collection.
+    @discardableResult
+    public mutating func shift<R : RangeExpression>(
+        from range: R, toJustBefore insertionPoint: Index
+    ) -> Range<Index> where R.Bound == Index
+
+    /// Moves the element at the given index to just before the specified
+    /// insertion point.
+    ///
+    /// This method moves the element at position `i` to immediately before
+    /// `insertionPoint`. This example shows moving elements forward and
+    /// backward in an array of integers.
+    ///
+    ///     var numbers = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+    ///     let newIndexOfNine = numbers.shift(from: 9, toJustBefore: 7)
+    ///     // numbers == [0, 1, 2, 3, 4, 5, 6, 9, 7, 8, 10]
+    ///     // newIndexOfNine == 7
+    ///
+    ///     let newIndexOfOne = numbers.shift(from: 1, toJustBefore: 4)
+    ///     // numbers == [0, 2, 3, 1, 4, 5, 6, 9, 7, 8, 10]
+    ///     // newIndexOfOne == 3
+    ///
+    /// To move an element to the end of a collection, pass the collection's
+    /// `endIndex` as `insertionPoint`.
+    ///
+    ///     numbers.shift(from: 0, toJustBefore: numbers.endIndex)
+    ///     // numbers == [2, 3, 1, 4, 5, 6, 9, 7, 8, 10, 0]
+    ///
+    /// - Parameters:
+    ///   - source: The index of the element to move. `source` must be a valid
+    ///     index of the collection that isn't `endIndex`.
+    ///   - insertionPoint: The index to use as the destination of the element.
+    ///     `insertionPoint` must be a valid index of the collection.
+    /// - Returns: The resulting index of the element that began at `source`.
+    ///
+    /// - Complexity: O(*n*) where *n* is the length of the collection.
+    @discardableResult
+    public mutating func shift(
+        from source: Index, toJustBefore insertionPoint: Index
+    ) -> Index
+}
+```
+
+#### Removing elements
+
+Within a range-replaceable collection, you can remove the elements represented by a range set. The implementation provides an additional, in-place overload for range-replaceable collections that also conform to `MutableCollection`.
+
+```swift
+extension RangeReplaceableCollection {
+    /// Removes the elements at the given indices.
+    ///
+    /// For example, this code sample finds the indices of all the vowel
+    /// characters in the string, and then removes those characters.
+    ///
+    ///     var str = "The rain in Spain stays mainly in the plain."
+    ///     let vowels: Set<Character> = ["a", "e", "i", "o", "u"]
+    ///     let vowelIndices = str.indices(where: { vowels.contains($0) })
+    ///
+    ///     str.removeAll(at: vowelIndices)
+    ///     // str == "Th rn n Spn stys mnly n th pln."
+    ///
+    /// - Parameter indices: The indices of the elements to remove.
+    ///
+    /// - Complexity: O(*n*), where *n* is the length of the collection.
+    public mutating func removeAll(at indices: RangeSet<Index>)
```

Comment:
This should be called `remove`, not `removeAll`.  Did we already add `removeAll(where:)`?  Oh, we did.  But still, the `All` isn't adding anything here.  In the case of `removeAll(where:)` it at least adds some fluency.

Reply:
Keeping this as `removeAll` — `remove(at:)` gets rid of a single element, while the multi-remove methods have a word indicating their multi-ness.